classes/task/collect_metrics_task.php

```diff
@@ -125,42 +79,36 @@ public function get_frequency_cutoff(int $timediff): int {
     public function execute() {
         // This algorithm is performance important. Any opportunity to optimize should be welcome.
 
-        // Use the server's timezone for determining times from strings.
-        $tz = \core_date::get_server_timezone_object();
-
-        // Get the time reference. The time string is in the server's timezone.
-        // It is then converted to a timestamp, which is UTC.
-        $reftime = (new \DateTime('midnight last Sunday', $tz))->getTimestamp();
-
-        // We don't need to use $tz here because timestamps are always UTC.
-        $time = $this->time ?? time();
+        $nowtimestamp = !is_null($this->time) ? $this->time : time();
 
-        // Get the number of minutes (rounded) that have passed since the ref time.
-        $timediff = (int) round(($time - $reftime) / MINSECS);
+        $metrictypes = metric\manager::get_metrics(true);
+        $items = [];
+        foreach ($metrictypes as $metrictype) {
+            if (!$metrictype->is_ready()) {
+                continue;
+            }
 
-        // Get the highest metric frequency to be record.
-        $cutoff = $this->get_frequency_cutoff($timediff);
+            $freq = $metrictype->get_frequency();
 
-        // We need to determine the monthly metrics separately, since a month is not a whole number of weeks.
-        $halfmin = MINSECS / 2;
-        $ismonth = 0;
-        if ($cutoff >= metric\manager::FREQ_DAY) {
-            $reftime = (new \DateTime('midnight this month -'. $halfmin .' seconds', $tz))->getTimestamp();
-            // Time difference should be zero if we are indeed at monthly boundary, but we allow for up to +/-30 seconds.
-            if (($time - $reftime) < MINSECS) {
-                $ismonth = metric\manager::FREQ_MONTH;
+            // When is the next generation due?
+            $lasttimestamp = $metrictype->get_last_generate_time();
+            if ($lasttimestamp === 0) {
+                // We generate the metric item regardless.
+                $duetimestamp = lib::get_last_whole_tick($nowtimestamp, $freq);
+                $lasttimestamp = lib::get_previous_time($duetimestamp, $freq);
+            } else {
+                $duetimestamp = lib::get_next_time(lib::get_last_whole_tick($lasttimestamp, $freq), $freq);
             }
-        }
 
-        $metrics = metric\manager::get_metrics(true);
-        $items = [];
-        foreach ($metrics as $metric) {
-            if (($metric->get_frequency() <= $cutoff || $metric->get_frequency() == $ismonth) && $metric->is_ready()) {
-                $items[] = $metric->get_metric_item();
+            if ($duetimestamp <= $nowtimestamp) {
+                mtrace('generating metric for ' . $metrictype->get_name() . ' from ' . userdate($lasttimestamp) . ' to ' . userdate($duetimestamp));
```

Comment:
1) This is looking nice - can we do this in a slightly different place so we can include the metric itself?

2) Also can use do some sprintf alignment so it looks like a text table.

3) tighten up the days and months just to make it a little shorter

```
generating metric for activeusers    13 from Wed, 1 Jun 2022, 11:53 AM to Wed, 1 Jun 2022, 11:54 AM
generating metric for newusers        6 from Wed, 1 Jun 2022, 11:53 AM to Wed, 1 Jun 2022, 11:54 AM
generating metric for onlineusers     8 from Wed, 1 Jun 2022, 11:53 AM to Wed, 1 Jun 2022, 11:54 AM

```